### PR TITLE
Collection avus, avu purges on deleted data objects (4-2-stable)

### PR DESCRIPTION
**Pull request by d-w-moore**

### CODE REVIEW

File: libirods_rule_engine_plugin-indexing.cpp [hunk 1]
```diff
@@ -288,23 +301,54 @@ namespace {
...
+                namespace fsvr = irods::experimental::filesystem::server;
+                fs::path data_logical_path {obj_inp->objPath};
+                std::string query_str {
+                    boost::str(boost::format("select DATA_ID,META_DATA_ATTR_NAME,META_DATA_ATTR_VALUE,META_DATA_ATTR_UNITS"
+                                             " where DATA_NAME = '%s' and COLL_NAME = '%s'"
+                                             ) % data_logical_path.object_name().string() 
+                                               % data_logical_path.parent_path().string() )
+                };
+                irods::query<rsComm_t> qobj{_rei->rsComm, query_str};
+                for (auto row : qobj) {
```
Comment (korydraughn) [inline libirods_rule_engine_plugin-indexing.cpp:341]: is this supposed to be a copy?

File: libirods_rule_engine_plugin-indexing.cpp [hunk 2]
```diff
@@ -319,6 +363,77 @@ namespace {
...
+                }
+                CollInp*obj_inp = nullptr;
+                obj_inp = boost::any_cast<CollInp*>(*it);
+                std::string query_str {
+                    boost::str(boost::format("select COLL_ID,COLL_NAME,META_COLL_ATTR_NAME,META_COLL_ATTR_VALUE,META_COLL_ATTR_UNITS"
+                                             " where COLL_NAME = '%s' || like '%s/%%'") % obj_inp->collName % obj_inp->collName )
+                };
+                irods::query<rsComm_t> c_qobj{_rei->rsComm, query_str};
+                auto &entry = avus_to_purge[obj_inp->collName];
+                for (auto row : c_qobj) {
```
Comment (korydraughn) [inline libirods_rule_engine_plugin-indexing.cpp:397]: is this supposed to be a copy?

File: libirods_rule_engine_plugin-indexing.cpp [hunk 3]
```diff
@@ -319,6 +363,77 @@ namespace {
...
+                        entry.emplace_back( row[0], row[1], row[2] , row[3] , row[4] );
+                }
+                std::string d_query_str {
+                    boost::str(boost::format("select DATA_ID,COLL_NAME,META_DATA_ATTR_NAME,META_DATA_ATTR_VALUE,META_DATA_ATTR_UNITS,DATA_NAME"
+                                             " where COLL_NAME = '%s' || like '%s/%%'") % obj_inp->collName % obj_inp->collName )
+                };
+                irods::query<rsComm_t> d_qobj{_rei->rsComm, d_query_str};
+                for (auto row : d_qobj) {
```
Comment (korydraughn) [inline libirods_rule_engine_plugin-indexing.cpp:405]: is this supposed to be a copy?

File: libirods_rule_engine_plugin-indexing.cpp (hunk 1 above)
Comment (d-w-moore) [reply libirods_rule_engine_plugin-indexing.cpp:341]: It doesn't have to be a copy. Will try as const reference

**Comment by trel** [comment]
so if you're happy, please squash to a commit or three that solve the enumerated issues.   we'll confirm everything looks good, and then later add the #s.

**Comment by d-w-moore** [comment]
Will do that soon; working on #19  now which I reproduced yesterday.

**Comment by d-w-moore** [comment]
All squashing and numbering is done.

File: libirods_rule_engine_plugin-elasticsearch.cpp [hunk 4]
```diff
@@ -36,6 +39,146 @@
 #include <algorithm>
 
 namespace {
+
+    using string_t = std::string;
+
+    // -- begin stolen code from old nlohmann::json
```
Comment (trel) [inline libirods_rule_engine_plugin-elasticsearch.cpp:45]: how about "begin code from"

File: libirods_rule_engine_plugin-elasticsearch.cpp [hunk 5]
```diff
@@ -36,6 +39,146 @@
...
+                    }
+                    break;
+                }
+            }
+        }
+
+        return result;
+    }
+
+    // --  end  stolen code from old nlohmann::json
```
Comment (trel) [inline libirods_rule_engine_plugin-elasticsearch.cpp:180]: same -  it is MIT licensed.

**Comment by trel** [comment]
very good - if you're happy with the state of this PR, please # and we'll get it merged.

**Pull request merged**